nixos/keter: init at 2.0.1

jappeace opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Adds keter as a module to nixos.
Currently keter is unusable with nix, 
because it relies on bundeling of a tar and uploading that to a specific folder.
These expressions automate this,
with especially nixops in mind.
This will work with more versions than just 2.0.1

I'm the maintainer of keter,
This module is pretty much as I've defined it in the [keter repo](https://github.com/snoyberg/keter#nixos).
Feedback is appreciated.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Lassulus commented:
hi, new modules would be nice to come with a test, so people can check if stuff breaks when updating :)

jappeace commented:
@Lassulus I added a test that runs keter

Ericson2314 reviewed (approved):
LGMT though I usually do not review these things

jappeace committed:
nixos/keter: init

Enable keter module

Keter is an apploader which:
1. has the old app running on a port.
2. loads a new one, and wait for that to complete
3. switches the old with the new one once the new one finished loading.

It supports more functionality but this use case
is the primary one being used by supercede.

Adds keter as a module to nixos.
Currently keter is unusable with nix,
because it relies on bundeling of a tar and uploading that to a specific folder.
These expressions automate these devops tasks,
with especially nixops in mind.
This will work with versions above 1.8

The test seems to work.
This uses a new version of keter which has good
support for status code on error pages.
We're using this config at production at supercede
so it should be fine.

Squash log:
==========

mention keter in changelog

Update generated release notes

Always restart keter on failure

This is a little bit of extra stability in case keter crashes.
Which can happen under extreme conditions (DoS attacks).

Update nixos/doc/manual/release-notes/rl-2205.section.md

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>

Update nixos/modules/module-list.nix

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>

Remove sanitization

don't put domain in as a string

Update nixos/tests/keter.nix

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>

add jappie as module maintainer

Use type path instead of two seperate options

Fix generated docs

added test machinery to figure out why it's failing

Fix the test, use console output

run nixpkgs-fmt on all modules

Inline config file.

This get's rid of a lot of inderection as well.

Run nix format

remove comment

simplify executable for test

delete config file

add config for keter root

Remove after redis clause

set keter root by default to /var/lib/keter

Update nixos/modules/services/web-servers/keter/default.nix

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>

Update nixos/modules/services/web-servers/keter/default.nix

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>

Update nixos/modules/services/web-servers/keter/default.nix

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>

fix nit

add newlines

add default text and move description in a long description

Delete rather obvious comment

fix release db thing

remove longDescription and put it in a comment instead

change description of mkEnalbeOption

explain what keter does by using the hackage synopsis

set domain to keterDomain and same for executable

move comment to where it's happening

fix type error

add formatting better comment

try add seperate user for keter

Revert "try add seperate user for keter"

This reverts commit d3522d36c96117335bfa072e6f453406c244e940.

Doing this breaks the setup

set default to avoid needing cap_net_bind_service

remove weird comment

use example fields

eleborated on process leakage

Update nixos/modules/services/web-servers/keter/default.nix

Co-authored-by: ckie <25263210+ckiee@users.noreply.github.com>

run nixpkgs-fmt

update docs

Fix formatting, set keter package by default

format our little nixexpr

replace '' -> " where possible

drop indent for multiline string

make description much shorter

regen docs database

ckiee commented on nixos/doc/manual/release-notes/rl-2205.section.md:
````diff
@@ -714,6 +714,7 @@ In addition to numerous new and upgraded packages, this release has the followin
   redis-cli save
   cp /var/lib/redis/dump.rdb "/var/lib/redis-mastodon/dump.rdb"
   ```
+- Added the `keter` module. This makes working with keter a lot easier for nixos and nix applications deployments.
````
```suggestion
- Added the `keter` NixOS module. This makes working with keter a lot easier for NixOS and Nix deployments.
```

ckiee commented on nixos/modules/services/web-servers/keter/bundle.nix:
```diff
@@ -0,0 +1,35 @@
+  };
+
+in stdenv.mkDerivation {
+  name = lib.strings.sanitizeDerivationName "keter-bundle";
```
It looks like a trivial/safe name to me, why sanitize?

jappeace replied:
sorry, this must have leaked trough all my copy paste editing. I shall remove the sanitization

ckiee commented on nixos/modules/services/web-servers/keter/config.nix:
```diff
@@ -0,0 +1,74 @@
+    wantedBy = [ "multi-user.target" "nginx.service" ];
+
+    serviceConfig = {
+      Restart="always";
+      RestartSec="10s";
```
Run `nixpkgs-fmt`

ckiee commented on nixos/modules/services/web-servers/keter/bundle.nix:
```diff
@@ -0,0 +1,35 @@
+    # we just use nix as an absolute path so we're not bundling any binaries
+    type = "webapp";
+    exec = "${exePkg}/bin/${executable}";
+    host = "${domain}";
```
```suggestion
    host = domain;
```

ckiee commented on nixos/modules/services/web-servers/keter/config.nix:
```diff
@@ -0,0 +1,74 @@
+    (bundleCfg // { exePkg = executable; });
+
+  executable = writeShellScriptBin "${bundleCfg.executable}" ''
+    #!/usr/bin/env bash
```
A shebang will be added for you by `writeShellScriptBin`, this is a noop

jappeace replied:
I deleted this

ckiee commented on nixos/modules/services/web-servers/keter/config.nix:
```diff
@@ -0,0 +1,74 @@
+
+  executable = writeShellScriptBin "${bundleCfg.executable}" ''
+    #!/usr/bin/env bash
+    export PATH=$PATH:${lib.makeBinPath [ coreutils ]}
```
Why do you need coreutils? You're only using shell builtins.

jappeace replied:
I deleted this

ckiee commented on nixos/modules/services/web-servers/keter/config.nix:
```diff
@@ -0,0 +1,74 @@
+  '';
+
+in {
+  # This is keter, which loads our app on deploy, and checks if it is healthy
+  # be very careful putting anything in here that changes,
+  # any change to the systemd unit can cause brief windows of downtime
+  # because nixos has to restart the unit.
```
Why is this here? Users might not look at this file before using it.

jappeace replied:
Where should I put this?

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,105 @@
+# https://nixos.wiki/wiki/Module
```
Comment seems redundant. Also you should run `nixpkgs-fmt` on this file too.

jappeace replied:
I think I formatted all the files

ckiee replied:
Comment still here

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,105 @@
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.keter = keterConfig.keter;
+    systemd.services.keter-load-bundle = keterConfig.load-keter-bundle;
```
If you're reading `config.nix` it's unclear that it's two systemd services, please inline that here or move the `systemd.services.*` part to `config.nix` too

jappeace replied:
I inlined the config file in here, this get's rid of a lot of inderection as well :+1:

ckiee replied:
I was thinking GitHub broke for me, but it's not inlined? Maybe git human error? Forgot to push?

jappeace replied:
hehe, yes, turns out airports aren't great for pushing so it didn't end up online even though I made the comment :sweat_smile:

ckiee commented on nixos/tests/keter.nix:
```diff
@@ -0,0 +1,32 @@
+import ./make-test-python.nix ({ pkgs, ... }: {
+  name = "keter";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ jappie ];
```
You need to add yourself to `/maintainers/maintainer-list.nix`.

jappeace replied:
I'm already in on line 5691, I maintain cut-the-crap as well

ckiee replied:
Huh, I searched for `japple` instead of `jappie` (^:

ckiee commented on nixos/tests/keter.nix:
```diff
@@ -0,0 +1,32 @@
+        bundle = {
+            domain = "example.com";
+            package = pkgs.writeShellScriptBin "run" ''
+              #! /bin/sh
```
```suggestion
```

ckiee commented on nixos/modules/services/web-servers/keter/bundle.nix:
```diff
@@ -0,0 +1,35 @@
+  str.stanzas = [{
+    # we just use nix as an absolute path so we're not bundling any binaries
+    type = "webapp";
+    exec = "${exePkg}/bin/${executable}";
```
Please just take in one `types.path` option instead of two options for the same thing.

jappeace replied:
I did this and removed the default, took me a while to figure out how to get the tests to work because the new keter version has better error handling.

ckiee commented on nixos/tests/keter.nix:
```diff
@@ -0,0 +1,32 @@
+              #! /bin/sh
+              ${pkgs.python3}/bin/python -m http.server $PORT
+            '';
+            executable = "run";
```
Will need to be changed according to one of my earlier comments, a little reminder.

jappeace replied:
yes this is a derivation now.

ckiee commented:
Commit message should be `nixos/keter: init`, you can change that with `git commit --amend`

jappeace commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,105 @@
+    keter = cfg.keterPackage;
+  };
+in
+{
```
added myself as maintainer

jappeace replied:
I added this as well

jappeace commented:
I also tested these changes on staging, seems to work, thanks for the review!

jappeace commented:
@ckiee please review again if you have time, I hope it's satisfactory now :crossed_fingers:

jappeace commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,105 @@
+
+    bundle = {
+        executable = lib.mkOption {
+            type = lib.types.str;
```
I made this a path

ckiee commented on nixos/tests/keter.nix:
```diff
@@ -0,0 +1,47 @@
+        bundle = {
+          appName = "test-bundle";
+          domain = "localhost";
+          executable =
+            let
+              exe = pkgs.writeShellScript "run" ''
+                ${pkgs.python3}/bin/python -m http.server $PORT
+              '';
+            in
+            "${exe}";
```
Is there really a reason you have to wrap this? What breaks?

(Also, this could be a one-liner: `executable = "${pkgs.writeShellScript "run" "${pkgs.python3}/bin/blahblahyoucanfillintherest"}"`

jappeace replied:
there was indeed no need to wrap this.
I didn't make it a 1 liner because that would make it go over 80 char limit.

ckiee commented on nixos/modules/services/web-servers/keter/config.nix:
```diff
@@ -0,0 +1,87 @@
+  # Furthermore, it's important that we use exec to
+  # run the binary otherwise we get process leakage.
+  executable = writeShellScript "bundle-wrapper" ''
+    #!/usr/bin/env bash
+    export PATH=$PATH:${lib.makeBinPath [ coreutils ]}
```
Still here

jappeace replied:
this was part of the git human/machine error

ckiee commented on nixos/modules/services/web-servers/keter/config.nix:
```diff
@@ -0,0 +1,87 @@
+    script = ''
+      set -xe
+      mkdir -p ${incoming}
+      { tail -F /opt/keter/log/keter/current.log -n 0 & ${keter}/bin/keter ${globalKeterConfigFile}; }
```
Why `/opt`? This is (almost) not a thing in NixOS.

Also, does `tail` die with the rest of the service? Check `journalctl` as it might be complaining about that on service restart.

jappeace replied:
good catch, that `/opt` should be keterRoot instead, it's by default `/opt/keter` because that's how snoyman described it in the readme before I took over as maintainer.
Perhaps there would be a better default for nixos?

--- 

about killing, here I'm `journalctl -fu keter`, the first two times I used systemctl restart, the last time I used pkill keter (to simulate a a crash), looks like it's just functioning as expected, it reboots with no complaints about tail as far as I can tell.

```
May 25 09:43:30 supercede_staging systemd[1]: Stopping keter app loader...
May 25 09:43:30 supercede_staging systemd[1]: keter.service: Deactivated successfully.
May 25 09:43:30 supercede_staging systemd[1]: Stopped keter app loader.
May 25 09:43:30 supercede_staging systemd[1]: keter.service: Consumed 6h 39min 1.650s CPU time, received 795.8M IP traffic, sent 876.6M IP traffic.
May 25 09:43:30 supercede_staging systemd[1]: Started keter app loader.
May 25 09:43:30 supercede_staging keter-start[3922060]: + mkdir -p /opt/keter/incoming
May 25 09:43:30 supercede_staging keter-start[3922060]: + /nix/store/pgrcw6dkpk8d3lwanby07xk6m2gx956c-keter-2.0.1/bin/keter /nix/store/25dmvgdl838a5dlq29jxq1cnkjdgr8ff-keter-config.yml
May 25 09:43:30 supercede_staging keter-start[3922062]: + tail -F /opt/keter/log/keter/current.log -n 0
May 25 09:43:30 supercede_staging keter-start[3922062]: tail: '/opt/keter/log/keter/current.log' has become inaccessible: No such file or directory
May 25 09:43:30 supercede_staging keter-start[3922062]: tail: '/opt/keter/log/keter/current.log' has appeared;  following new file
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Launching cli
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Launching initial
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/myapp.keter\" 1653300542"
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Started watching
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Unpacking bundle '/opt/keter/incoming/myapp.keter'
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/supercede.keter\" 1653410590"
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Started listening
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Unpacking bundle '/opt/keter/incoming/supercede.keter'
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Sanity checks passed
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Sanity checks passed
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Reserving hosts for app myapp: staging.supercede.com
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Reserving hosts for app supercede: staging.supercede.com
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Process created: /nix/store/73hjl15h722fxbdiap547mxzrsy2nalp-bundle-wrapper
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Error occured when launching bundle supercede: CannotReserveHosts supercede (fromList [("staging.supercede.com",myapp)])
May 25 09:43:34 supercede_staging keter-start[3922062]: 2022-05-25 09:43:34.20: Activating app myapp with hosts: staging.supercede.com
May 25 09:43:59 supercede_staging systemd[1]: Stopping keter app loader...
May 25 09:43:59 supercede_staging systemd[1]: keter.service: Deactivated successfully.
May 25 09:43:59 supercede_staging systemd[1]: Stopped keter app loader.
May 25 09:43:59 supercede_staging systemd[1]: keter.service: Consumed 16.132s CPU time, received 528.3K IP traffic, sent 673.9K IP traffic.
May 25 09:43:59 supercede_staging systemd[1]: Started keter app loader.
May 25 09:43:59 supercede_staging keter-start[3922122]: + mkdir -p /opt/keter/incoming
May 25 09:43:59 supercede_staging keter-start[3922122]: + /nix/store/pgrcw6dkpk8d3lwanby07xk6m2gx956c-keter-2.0.1/bin/keter /nix/store/25dmvgdl838a5dlq29jxq1cnkjdgr8ff-keter-config.yml
May 25 09:43:59 supercede_staging keter-start[3922124]: + tail -F /opt/keter/log/keter/current.log -n 0
May 25 09:43:59 supercede_staging keter-start[3922124]: tail: '/opt/keter/log/keter/current.log' has become inaccessible: No such file or directory
May 25 09:43:59 supercede_staging keter-start[3922124]: tail: '/opt/keter/log/keter/current.log' has appeared;  following new file
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Launching cli
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Launching initial
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/myapp.keter\" 1653300542"
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Started watching
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Unpacking bundle '/opt/keter/incoming/myapp.keter'
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/supercede.keter\" 1653410590"
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Unpacking bundle '/opt/keter/incoming/supercede.keter'
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Started listening
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Sanity checks passed
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Reserving hosts for app myapp: staging.supercede.com
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Sanity checks passed
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Reserving hosts for app supercede: staging.supercede.com
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Error occured when launching bundle supercede: CannotReserveHosts supercede (fromList [("staging.supercede.com",myapp)])
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Process created: /nix/store/73hjl15h722fxbdiap547mxzrsy2nalp-bundle-wrapper
May 25 09:44:03 supercede_staging keter-start[3922124]: 2022-05-25 09:44:03.68: Activating app myapp with hosts: staging.supercede.com
May 25 09:44:54 supercede_staging systemd[1]: keter.service: Deactivated successfully.
May 25 09:44:54 supercede_staging systemd[1]: keter.service: Consumed 21.838s CPU time, received 798.0K IP traffic, sent 985.2K IP traffic.
May 25 09:45:04 supercede_staging systemd[1]: keter.service: Scheduled restart job, restart counter is at 1.
May 25 09:45:04 supercede_staging systemd[1]: Stopped keter app loader.
May 25 09:45:04 supercede_staging systemd[1]: keter.service: Consumed 21.838s CPU time, received 798.0K IP traffic, sent 985.2K IP traffic.
May 25 09:45:04 supercede_staging systemd[1]: Started keter app loader.
May 25 09:45:04 supercede_staging keter-start[3922208]: + mkdir -p /opt/keter/incoming
May 25 09:45:04 supercede_staging keter-start[3922208]: + /nix/store/pgrcw6dkpk8d3lwanby07xk6m2gx956c-keter-2.0.1/bin/keter /nix/store/25dmvgdl838a5dlq29jxq1cnkjdgr8ff-keter-config.yml
May 25 09:45:04 supercede_staging keter-start[3922210]: + tail -F /opt/keter/log/keter/current.log -n 0
May 25 09:45:04 supercede_staging keter-start[3922210]: tail: '/opt/keter/log/keter/current.log' has been replaced;  following new file
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Launching cli
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Launching initial
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/myapp.keter\" 1653300542"
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Unpacking bundle '/opt/keter/incoming/myapp.keter'
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Started watching
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/supercede.keter\" 1653410590"
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Sanity checks passed
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Unpacking bundle '/opt/keter/incoming/supercede.keter'
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Reserving hosts for app myapp: staging.supercede.com
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Started listening
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Sanity checks passed
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Process created: /nix/store/73hjl15h722fxbdiap547mxzrsy2nalp-bundle-wrapper
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Reserving hosts for app supercede: staging.supercede.com
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Error occured when launching bundle supercede: CannotReserveHosts supercede (fromList [("staging.supercede.com",myapp)])
May 25 09:45:08 supercede_staging keter-start[3922210]: 2022-05-25 09:45:08.25: Activating app myapp with hosts: staging.supercede.com
```

ckiee replied:
`/var/lib/keter` seems to be the consensus, on my current machine:
```
ckie@cookiemonster ~ -> ls /var/lib
AccountsService   alsa  cntr        cups      dnscrypt-proxy  lightdm       logrotate.status  misc            NetworkManager-fortisslvpn  private  tailscale
adguard-dnsproxy  cni   containers  dhclient  libvirt         lightdm-data  machines          NetworkManager  nixos                       systemd  udisks2
ckie@cookiemonster ~ -> ls /opt
"/opt": No such file or directory (os error 2)
```

ckiee commented on nixos/modules/services/web-servers/keter/config.nix:
```diff
@@ -0,0 +1,87 @@
+      "network.target"
+      "local-fs.target"
+      "postgresql.service"
+      "redis.service"
```
Which Redis? The Redis NixOS module supports multiple instances, go look at that, and add an instance to this module.

jappeace replied:
good point, the app being loaded may need redis, but keter itself doesn't, I removed it for now

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+        type = lib.types.str;
+        default = "myapp";
+        description = ''The name keter assigns to this bundle'';
+      };
+      executable = lib.mkOption {
```
Nit: no space between these two

ckiee commented on nixos/tests/keter.nix:
```diff
@@ -0,0 +1,43 @@
+          domain = "localhost";
+          executable = pkgs.writeShellScript "run" ''
+                ${pkgs.python3}/bin/python -m http.server $PORT
+              '';
```
Fix indent here to be on the same level as `executable =`

ckiee commented on nixos/modules/services/web-servers/keter/bundle.nix:
```diff
@@ -0,0 +1,41 @@
+let
+  str.stanzas = [{
+    # we just use nix as an absolute path so we're not bundling any binaries
+    type = "webapp";
```
I don't know much about keter, but what if someone wants a different configuration? You should expose this `str` thing. (And maybe name it better?)

jappeace replied:
there are several options available: https://github.com/snoyberg/keter/blob/master/src/Keter/Config/V10.hs#L224
but I'm not sure if it's a good idea to support these.
We only use the webapp configuration at supercede because it solves a relatively hard problem of app loading:
Keep old app running, swap the app ports once the new one has finsihed booting

but the other options allow you to configure keter as an nginx replacement (with static-files or reverse-proxy), and as a background task manager.
To do that we'd need to make the rest of the stanza modifiable as well and I'm not sure if it's worth the effort.

ckiee replied:
Can you change the `mkEnableOption` then? [^1]
```nix
    enable = lib.mkEnableOption "keter, a web app deployment manager. Note this module only supports application hotswap or whatever this is called you change this";
 ```
[^1]: Also, having two instances of your app, load balancing between them and A-B updates seems like a saner solution but that's beside the point.

jappeace replied:
I changed the description like you suggested.
Not sure what you're getting at with loadbalancing. we don't need that, a single vm will be able to serve the entire reinsurance market.

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+    maintainers = [ jappie ];
+  };
+  options.services.keter = {
+    enable = lib.mkEnableOption "Keter";
```
```suggestion
    enable = lib.mkEnableOption "keter, a web app deployment manager";
```

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+{
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ jappie ];
+  };
+  options.services.keter = {
```
Nit: no newline between these two

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+    keterPackage = lib.mkOption {
+      type = lib.types.package;
+      default = null;
+      description = ''set the keter package to be used, usually pkgs.haskellPackages.keter'';
```
```suggestion
      description = ''The keter package to be used, usually <varname>pkgs.haskellPackages.keter</varname>'';
```

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+    keterRoot = lib.mkOption {
+      type = lib.types.str;
+      default = "/var/lib/keter";
+      description = ''Mutable state folder for keter. In here the bundles get extracted,
+                      logs get written and the app's work folder is set to'';
```
```suggestion
      description = ''Mutable state folder for keter'';
```

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+    };
+    keterPackage = lib.mkOption {
+      type = lib.types.package;
+      default = null;
```
This is a `lib.types.package` but you are passing `null`. Explain?

jappeace replied:
It needs to be set by the user, I couldn't figure out how to set this by default to `haskellPackges.keter`

ckiee replied:
What does it do when you set it to `pkgs.haskellPackges.keter`? (And, if it works, also set defaultText ([example](https://github.com/nixos/nixpkgs/blob/3939c4bb8caf6458d1f4b2f1bacc221503d12afa/nixos/modules/config/users-groups.nix#L174)))

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+          port = 81;
+        }];
+      };
+      description = ''root is required to function in nixos .
+                      port 81 is set to default to allow nginx to reverse proxy into keter,
+                      so that nginx can do ssl management.
+                      You want that ip-from-header in the nginx setup case
+                      so it's not set to 127.0.0.1.
+                      Keep in mind root is overwritten by keterRoot.
+                      which is required for the nix derivation.
+                      '';
```
Just say something like "Global configuration for keter", maybe add a `longDescription` and definitely add a `defaultText = lib.literalExpression "[{ port = 81; TODO fill this in }]"`

jappeace replied:
I did this

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+
+    in
+    {
+      # This is keter, which loads our app on deploy, and checks if it is healthy
+      # be very careful putting anything in here that changes,
+      # any change to the systemd unit can cause brief windows of downtime
+      # because nixos has to restart the unit.
```
Like what? If the keter package gets updated that seems natural to me that it'd restart it.

jappeace replied:
Yes.. you'd say so...
this was an internal comment disgurgaging against merging an upgraded nixpkgs and large migrations in a single deploy cycle.
I deleted it.

ckiee reviewed (changes requested):
Need to rerun `nixpkgs-fmt` first before addressing other comments. Also please mark things as resolved when you've done them.

(Somehow I managed to do all of this in 16 minutes, not including writing this, so I think that's pretty neat `(^:`)

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,166 @@
+  cfg = config.services.keter;
+in
+{
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ jappie ];
+  };
```
```suggestion
  meta = {
    maintainers = with lib.maintainers; [ jappie ];
  };
```

ckiee reviewed (changes requested):
Haven't fully gone over it again since the last review is still unaddressed. GitHub did a wonky and reposted the last review.

jappeace commented on nixos/modules/services/web-servers/keter/bundle.nix:
```diff
@@ -0,0 +1,41 @@
+/* This makes a keter bundle as described on the github page:
+  https://github.com/snoyberg/keter#bundling-your-app-for-keter
+
+  Note that we're not actually putting in the app into the bundle,
+  we already can use the nix store for copying, so we just
+  symlink to the app.
```
It's the configuration file being generated, I moved the comment into the stanzas

jappeace commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,166 @@
+          port = 81;
+        }];
+      };
+      # root is required to function in nixos .
```
I set the capability by default, I also tried making keter run under a different user but it broke the setup, besides I think keter has some support for forking out processes as a different user.

jappeace commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,166 @@
+      };
+      # root is required to function in nixos .
+      # port 81 is set to default to allow nginx to reverse proxy into keter,
+      # so that nginx can do ssl management.
+      # You want that ip-from-header in the nginx setup case
+      # so it's not set to 127.0.0.1.
+      # Keep in mind root is overwritten by keterRoot.
+      # which is required for the nix derivation.
```
You can also run keter without nginx, but at our company we did it with nginx because ssl was already setup for that.

jappeace commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,166 @@
+        description = ''The domain keter will bind to'';
+      };
+
+      # these won't work with standard nix options because they have
+      # to be put in a tarball, so we use a runtime script to load settings.
```
Maybe I meant passing it in as an environment variable to the derivation? The idea is that these get loaded and logged (hence public), 
I'm not sure why this wouldn't work with setting it in the executable either.
The secret script will use cat to load environment variables so they don't end up in the nix store, so I'm quite sure that's neccisary.

jappeace commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,166 @@
+        }];
+      };
+      # root is required to function in nixos .
+      # port 81 is set to default to allow nginx to reverse proxy into keter,
```
I set the default to 6981

jappeace commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,166 @@
+
+      # This indirection is required to ensure the nix path
+      # gets copied over to the target machine in remote deployments.
+      # Furthermore, it's important that we use exec to
+      # run the binary otherwise we get process leakage.
```
this get's executed every deploy so you end up with many more, we had some issues with this, which was worked around by just killing all the old webapps everytime cpu peaked above 80%.
Untill I finally found out this was the cause.

jappeace commented:
@ckiee please re-review :sweat_smile: 
I think I addressed everything (appologies if not, it becomes hard to keep track of what's resolved or not)

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,166 @@
+    keterRoot = lib.mkOption {
+      type = lib.types.str;
+      default = "/var/lib/keter";
+      description = ''Mutable state folder for keter'';
```
It's a `''` string even though it could be a `"` string.

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+    keterPackage = lib.mkOption {
+      type = lib.types.package;
+      default = null;
+      description = ''The keter package to be used, usually <varname>pkgs.haskellPackages.keter</varname>'';
```
No crossover from `defaultText` please.

jappeace replied:
I set the defaultText instead of varname, I also set the default to `pkgs.haskellPackages.keter` I'm not sure why I thought I had to set this to null before, it just works now.

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+      # so it's not set to 127.0.0.1.
+      # using a port above 1024 allows you to avoid needing CAP_NET_BIND_SERVICE
+      defaultText = lib.literalExpression ''
+        {ip-from-header = true; listeners = [{host = "*4"; port = 6981;}];}
```
Format our little nixexpr here? :cowboy_hat_face:

jappeace replied:
This is now formatted (copied from above)

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+      appName = lib.mkOption {
+        type = lib.types.str;
+        default = "myapp";
+        description = ''The name keter assigns to this bundle'';
+      };
+
+      executable = lib.mkOption {
+        type = lib.types.path;
+        description = ''The executable to be run'';
+      };
+
+      domain = lib.mkOption {
+        type = lib.types.str;
+        default = "example.com";
+        description = ''The domain keter will bind to'';
+      };
```
Same as before for all of these `''` strings, but..

(Turn these into `"this"`)

jappeace replied:
I `s/''/"` all single line ones.

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+      publicScript = lib.mkOption {
+        type = lib.types.str;
+        default = "";
+        description = ''
+          Allows loading of public environment variables,
+          these are emitted to the log so it shouldn't contain secrets.
+        '';
```
Not for this one, since it's actually multiline.

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
````diff
@@ -0,0 +1,163 @@
+        default = "";
+        description = ''
+          Allows loading of private environment variables,
+          ```
+          ```
````
?

jappeace replied:
Likely cut and pasted into the example, I deleted this.

ckiee commented on nixos/doc/manual/from_md/release-notes/rl-2205.section.xml:
```diff
@@ -2122,6 +2122,17 @@ sudo mkdir /var/lib/redis-peertube
 sudo cp /var/lib/redis/dump.rdb /var/lib/redis-peertube/dump.rdb
```
Commit should be called `nixos/keter: init`.

ckiee commented on nixos/doc/manual/release-notes/rl-2205.section.md:
````diff
@@ -776,6 +776,7 @@ In addition to numerous new and upgraded packages, this release has the followin
   sudo mkdir /var/lib/redis-peertube
   sudo cp /var/lib/redis/dump.rdb /var/lib/redis-peertube/dump.rdb
   ```
+- Added the `keter` NixOS module. Keter is an apploader, it Binds to the main port (usually port 80) and reverse proxies requests to your application based on virtual hostnames. It provides graceful redeployment support, by launching a second copy of your application, performing a health check, and then switching reverse proxying to the new process. 
````
Look around at [all of these](https://github.com/NixOS/nixpkgs/blob/ad8276cd21c9ef01e07e299ab20337722db5195a/nixos/doc/manual/release-notes/rl-2205.section.md#new-services-sec-release-2205-new-services). They're short, try to make your sentence smaller. (Drop the details)

jappeace replied:
I dropped the details

ckiee commented on nixos/modules/services/web-servers/keter/default.nix:
```diff
@@ -0,0 +1,163 @@
+
+  options.services.keter = {
+    enable = lib.mkEnableOption ''keter, a web app deployment manager.
+        Note that this module only support loading of webapps:
+        Keep an old app running and swap the ports when the new one is booted.
```
Why does this get extra indent? This will also get rendered to a manpage where space is not infinite. :P

jappeace replied:
I deleted the extra indent, I didn't realize this was used in manpages, that's pretty cool.

ckiee replied:
`man configuration.nix`

ckiee reviewed (changes requested):
Unrelated to the review but I don't understand why you need all of this instead of having two instances of your app behind a nginx load balancer to get all the same effects without all of the hassle keter brings. 

Getting pretty close to being done I think..

ckiee commented:
> please tell me more slightly_smiling_face
> I'm mainting keter for supercede but if I could replace it with
> a stock solution all the better.

Set up your NixOS config so each instance can take in it's own app derivation, then you can do an A/B style thing, separating your deploys into two:
- Switch A to new version
- Wait for it to start up, see that works
- Switch B to new version

<hr>

Anyway, back on the whole review grind thing, I'll reply to everything again soon. Hopefully not going to give up on this before you since you seem to have pretty incredible amounts of patience even as I pick apart every other line and point at the smallest things.

jappeace commented:
If by instance you mean a separate nixos virtualization, then we don't have that right now. Our setup is a single monolith nixos, and keter facilitates the switching. 
The A/B setup you mentioned would require some automation around deciding where to deploy to (the waiting), which is now handled by keter. (we'd also have to put the database on a separate machine as well and connect to that from the app, making things harder)
I think this keter setup is quite simple in terms of ops, 
and since keter is a fairly old piece of software we assumed it was production ready. Unlike making our own scripts for automated A/B deployment, which would require debugging etc.

----

>  Hopefully not going to give up on this before you since you seem to have pretty incredible amounts of patience even as I pick apart every other line and point at the smallest things.

Take your time, I take long breaks as well, I just come back to this once every so many weeks since it's on my list of things I need to upstream.

ckiee commented:
> If by instance you mean a separate nixos virtualization, then we don't have that right now. Our setup is a single monolith nixos, and keter facilitates the switching. The A/B setup you mentioned would require some automation around deciding where to deploy to (the waiting), which is now handled by keter. (we'd also have to put the database on a separate machine as well and connect to that from the app, making things harder) I think this keter setup is quite simple in terms of ops, and since keter is a fairly old piece of software we assumed it was production ready. Unlike making our own scripts for automated A/B deployment, which would require debugging etc.

Very fair points if you don't want the extra human error. It doesn't matter too much for my usecase so I didn't really think it through.
<hr>

> Take your time, I take long breaks as well, I just come back to this once every so many weeks since it's on my list of things I need to upstream.

My TODO list for free software is my inbox, so it's either now or never and I have one more PR to work on.. Let's go.

ckiee commented:
Just `nixpkgs-fmt`'ed the NixOS test file. Merging once Ofborg finishes.

jappeace commented:
Thanks!
